File: DesignPatternsExercises/java/BuilderDesignPattern/Solution/PCBuilder.java
// public __________(13)__________ class PCBuilder {
public abstract class PCBuilder {
    protected Computer computer;

    public PCBuilder() {
        // computer = __________(14)__________();
        computer = new Computer();
    }

    public PCBuilder addMonitor(int size) {
        // Monitor monitor = __________(15)__________(size);
        Monitor monitor = new Monitor(size);
        computer.addMonitor(monitor);
        // __________(16)__________;
        return this;
    }

    public PCBuilder addRam(int size) {
        // Ram ram = __________(17)__________(size);
        Ram ram = new Ram(size);
        computer.addRam(ram);
        // __________(18)__________;
        return this;
    }

    public PCBuilder addHardDisk(int size) {
        // HardDisk hardDisk = __________(19)__________(size);
        HardDisk hardDisk = new HardDisk(size);
        computer.addHardDisk(hardDisk);
        // __________(20)__________;
        return this;
    }

    public Computer getComputer() {
        return computer;
    }

    // public __________(21)__________ PCBuilder __________(22)__________(String type);
    public abstract PCBuilder addMouse(String type);
}

File: DesignPatternsExercises/ruby/builder_design_pattern/exercise/pc_builder.rb
require_relative 'computer'
require_relative 'hard_disk'
require_relative 'monitor'
require_relative 'ram'

class PCBuilder
  attr_reader __________(9)__________

  def __________(10)__________
    @computer = __________(11)__________
  end

  def add_monitor(size)
    monitor = __________(12)__________(size)
    @computer.add_monitor(monitor)
    __________(13)__________
  end

  def add_ram(size)
    ram = __________(14)__________(size)
    @computer.add_ram(ram)
    __________(15)__________
  end

  def add_hard_disk(size)
    hard_disk = __________(16)__________(size)
    @computer.add_hard_disk(hard_disk)
    __________(17)__________
  end

  def add_mouse(type)
    raise NotImplementedError, 'Subclasses must override this method'
  end
end
